coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala

```diff
@@ -0,0 +1,229 @@
+package filodb.coordinator
+
+import akka.actor.{Actor, ActorRef, Props}
+import com.typesafe.config.Config
+import net.ceedubs.ficus.Ficus._
+import org.velvia.filo.RowReader
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+import filodb.core._
+import filodb.core.Types._
+import filodb.core.metadata.{Column, Dataset, MetaStore, Projection, RichProjection}
+import filodb.core.columnstore.ColumnStore
+import filodb.core.reprojector.Reprojector
+
+/**
+ * The NodeCoordinatorActor is the common API entry point for all FiloDB ingestion and metadata operations.
+ * It is a singleton - there should be exactly one such actor per node/JVM process.
+ * It is responsible for:
+ * - Overall external FiloDB API.
+ * - Staying aware of status of other NodeCoordinators around the ring
+ * - Metadata changes (dataset/column changes)
+ * - Caching changes to dataset metadata?
+ * - Supervising, spinning up, cleaning up DatasetCoordinatorActors
+ * - Forwarding new changes (rows) to other NodeCoordinatorActors if they are not local
+ * - Forwarding rows to DatasetCoordinatorActors
+ *
+ * It is called by local (eg HTTP) as well as remote (eg Spark ETL) processes.
+ */
+object NodeCoordinatorActor {
+  /**
+   * Creates a new dataset with columns and a default projection.
+   */
+  case class CreateDataset(dataset: Dataset, columns: Seq[Column])
+
+  case object DatasetCreated extends Response
+  case class DatasetError(msg: String) extends ErrorResponse
+
+  /**
+   * Sets up ingestion for a given dataset, version, and schema of columns.
+   * The dataset and columns must have been previously defined.
+   *
+   * @defaultPartitionKey if Some(key), a null value in partitioning column will cause key to be used.
+   *                      if None, then NullPartitionValue will be thrown when null value
+   *                        is encountered in a partitioning column.
+   * @return BadSchema if the partition column is unsupported, sort column invalid, etc.
+   */
+  case class SetupIngestion(dataset: String,
+                            schema: Seq[String],
+                            version: Int,
+                            defaultPartitionKey: Option[PartitionKey] = None)
+
+  case object IngestionReady extends Response
+  case object UnknownDataset extends ErrorResponse
+  case class UndefinedColumns(undefined: Seq[String]) extends ErrorResponse
+  case class BadSchema(message: String) extends ErrorResponse
+
+  /**
+   * Ingests a new set of rows for a given dataset and version.
+   * The partitioning column and sort column are set up in the dataset.
+   *
+   * @param seqNo the sequence number to be returned for acknowledging the entire set of rows
+   * @return Ack(seqNo) returned when the set of rows has been committed to the MemTable.
+   */
+  case class IngestRows(dataset: String, version: Int, rows: Seq[RowReader], seqNo: Long)
+
+  case class Ack(seqNo: Long) extends Response
```

Comment:
I had all the responses extend NodeResponse as well.  The original `Response` was from earlier days, maybe get rid of it, though I like the idea of having a separate ErrorResponse thing.